File: Client/Services/StopAppointmentReference.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Client.Services;

public static partial class StopAppointmentReference
{
  private const int MaximumInput = 4096;
  private const int MaximumReferences = 4;

  public static IReadOnlyList<string> Extract(string? notes, string? job)
  {
    if (string.IsNullOrEmpty(notes)) return [];
    var input = notes[..Math.Min(notes.Length, MaximumInput)];
    var kind = Kind(job);
    var references = new List<string>();
    foreach (Match match in Reference().Matches(input))
    {
      var qualifier = match.Groups[1].Value;
      var value = match.Groups[2].Value;
      if (qualifier.Length > 0 && Kind(qualifier) != kind || !IsReference(value)
        || !EndsClause(notes, match.Index + match.Length, input.Length)) continue;
      if (!references.Contains(value, StringComparer.OrdinalIgnoreCase)) references.Add(value);
      if (references.Count == MaximumReferences) break;
    }
    return references;
  }

  private static string? Kind(string? value)
  {
    var normalized = new string((value ?? "").Where(char.IsLetter).ToArray()).ToLowerInvariant();
    var pickup = normalized.Contains("pickup", StringComparison.Ordinal) || normalized is "pu" or "shipper";
    var delivery = normalized.Contains("delivery", StringComparison.Ordinal) || normalized.Contains("dropoff", StringComparison.Ordinal)
      || normalized is "del" or "deliver" or "receiver";
    return pickup == delivery ? null : pickup ? "pickup" : "delivery";
  }

  private static bool EndsClause(string notes, int index, int limit)
  {
    if (index < limit && notes[index] is '\'' or '"') index++;
    while (index < limit && notes[index] is ' ' or '\t') index++;
    if (index == limit) return notes.Length == limit;
    if (notes[index] is not ('.' or ',' or ';' or '\r' or '\n')) return false;
    return notes[index] is not ('.' or ',') || index + 1 == notes.Length || !char.IsAsciiLetterOrDigit(notes[index + 1]);
  }

  private static bool IsReference(string value)
  {
    if (!value.Any(char.IsAsciiDigit) || DateOrOtherId().IsMatch(value)) return false;
    if (value.Length == 4 && int.TryParse(value, out var year) && year is >= 1900 and <= 2099) return false;
    if (value.Length != 8 || !value.All(char.IsAsciiDigit)) return true;
    foreach (var format in new[] { "yyyyMMdd", "MMddyyyy", "ddMMyyyy" })
      if (DateOnly.TryParseExact(value, format, CultureInfo.InvariantCulture, DateTimeStyles.None, out var date)
        && date.Year is >= 1900 and <= 2099) return false;
    return true;
  }

  // Only standalone labeled clauses are trusted; surrounding prose is not a fallback identifier source.
  [GeneratedRegex("(?:^|[.;,\\r\\n])\\s*(?:[-*•]\\s*)?(?:(shipper|pick\\s*up|receiver|deliver(?:y)?|drop\\s*off)(?:['’]s)?\\s*:?\\s*)?(?:appointment|appt)\\b\\s*(?:(?:confirmation|confirm|conf\\.?)\\s*(?:number|no\\.?|#)?|number|no\\.?|#)\\s*[:=#-]?\\s*[\"']?([A-Z0-9][A-Z0-9_/-]{2,63})", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant | RegexOptions.NonBacktracking)]
  private static partial Regex Reference();
  [GeneratedRegex(@"^(?:\d{1,4}[-/_]\d{1,2}[-/_]\d{1,4}|\d{1,2}[-/]\d{1,2}|\d{1,4}(?:am|pm)|(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)[a-z]*[-_]?\d{1,2}[-_]?\d{2,4}|\d{1,2}[-_]?(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)[a-z]*[-_]?\d{2,4}|(?:bol|load|order)[-_]?\d.*)$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant | RegexOptions.NonBacktracking)]
  private static partial Regex DateOrOtherId();
}

File: Client/Scripts/fleetMap/routes/stopAppointmentReference.ts
const maximumInput = 4096,
  maximumReferences = 4;
const reference =
  /(?:^|[.;,\r\n])\s*(?:[-*•]\s*)?(?:(shipper|pick\s*up|receiver|deliver(?:y)?|drop\s*off)(?:['’]s)?\s*:?\s*)?(?:appointment|appt)\b\s*(?:(?:confirmation|confirm|conf\.?)\s*(?:number|no\.?|#)?|number|no\.?|#)\s*[:=#-]?\s*["']?([A-Z0-9][A-Z0-9_/-]{2,63})/gi;
const dateOrOtherId =
  /^(?:\d{1,4}[-/_]\d{1,2}[-/_]\d{1,4}|\d{1,2}[-/]\d{1,2}|\d{1,4}(?:am|pm)|(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)[a-z]*[-_]?\d{1,2}[-_]?\d{2,4}|\d{1,2}[-_]?(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)[a-z]*[-_]?\d{2,4}|(?:bol|load|order)[-_]?\d.*)$/i;

function kind(value: unknown): 'pickup' | 'delivery' | null {
  const normalized = String(value ?? '')
    .replace(/[^a-z]/gi, '')
    .toLowerCase();
  const pickup =
    normalized.includes('pickup') || ['pu', 'shipper'].includes(normalized);
  const delivery =
    normalized.includes('delivery') ||
    normalized.includes('dropoff') ||
    ['del', 'deliver', 'receiver'].includes(normalized);
  return pickup === delivery ? null : pickup ? 'pickup' : 'delivery';
}

function endsClause(notes: string, index: number, limit: number): boolean {
  if (index < limit && ['"', "'"].includes(notes[index])) index++;
  while (index < limit && [' ', '\t'].includes(notes[index])) index++;
  if (index === limit) return notes.length === limit;
  if (!['.', ',', ';', '\r', '\n'].includes(notes[index])) return false;
  return (
    !['.', ','].includes(notes[index]) ||
    index + 1 === notes.length ||
    !/[a-z0-9]/i.test(notes[index + 1])
  );
}

function isReference(value: string): boolean {
  if (!/\d/.test(value) || dateOrOtherId.test(value)) return false;
  if (/^(?:19|20)\d{2}$/.test(value)) return false;
  if (!/^\d{8}$/.test(value)) return true;
  const formats = [
    [value.slice(0, 4), value.slice(4, 6), value.slice(6)],
    [value.slice(4), value.slice(0, 2), value.slice(2, 4)],
    [value.slice(4), value.slice(2, 4), value.slice(0, 2)],
  ];
  return !formats.some(parts => {
    const [year, month, day] = parts.map(Number),
      date = new Date(Date.UTC(year, month - 1, day));
    return (
      year >= 1900 &&
      year <= 2099 &&
      date.getUTCFullYear() === year &&
      date.getUTCMonth() === month - 1 &&
      date.getUTCDate() === day
    );
  });
}

// The appointment numbers written into a stop's notes, for the job that
// stop is.
export function stopAppointmentReference(
  notes: unknown,
  job: unknown,
): string[] {
  if (typeof notes !== 'string' || !notes.length) return [];
  const input = notes.slice(0, maximumInput),
    selectedKind = kind(job),
    values: string[] = [];
  for (const match of input.matchAll(reference)) {
    const [, qualifier, value] = match;
    if (
      (qualifier && kind(qualifier) !== selectedKind) ||
      !isReference(value) ||
      !endsClause(notes, match.index + match[0].length, input.length)
    )
      continue;
    if (
      !values.some(existing => existing.toLowerCase() === value.toLowerCase())
    )
      values.push(value);
    if (values.length === maximumReferences) break;
  }
  return values;
}
